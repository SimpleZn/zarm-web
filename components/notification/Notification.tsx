import React from 'react';
import classnames from 'classnames';
import Icon from '../icon';
import { NotificationProps, NotificationIcon } from './PropsType';
import { mapToIconType, mapToIconTheme, getStyle } from './utils';

function getIcon(icon: NotificationIcon | React.ReactElement, className: string) {
  if (React.isValidElement(icon)) {
    return (
      <div className={className}>{icon}</div>
    );
  }
  const iconType = mapToIconType(icon);
  return iconType
    ? (
      <Icon
        type={iconType}
        className={className}
        size="sm"
        theme={mapToIconTheme(icon)}
      />
    )
    : null;
}

export class Notification extends React.Component<NotificationProps, {}> {
  static defaultProps: NotificationProps = {
    prefixCls: 'zw-notification',
    content: '',
    title: '',
  };

  render() {
    const {
      style, className, prefixCls, title, icon, content, footer,
      top, bottom,
      onMouseEnter, onMouseLeave, onClose, onClick,
    } = this.props;
    const iconToRender = icon ? getIcon(icon, `${prefixCls}__icon`) : null;
    const cls = classnames(`${prefixCls}__content`, { 'has-icon': iconToRender });
    return (
      <div
        className={classnames(prefixCls, className, {
          [`${prefixCls}--fixed`]: top || bottom,
        })}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        style={getStyle(style, top, bottom)}
      >
        <div className={cls} onClick={onClick}>
          {iconToRender}
          {title && <div className={`${prefixCls}__head`}>{title}</div>}
          {content && <div className={`${prefixCls}__body`}>{content}</div>}
          {footer && <div className={`${prefixCls}__foot`}>{footer}</div>}
        </div>
        <Icon
          className={`${prefixCls}__close`}
          type="wrong"
          size="sm"
          onClick={onClose}
        />
      </div>
    );
  }
}

export default Notification;
