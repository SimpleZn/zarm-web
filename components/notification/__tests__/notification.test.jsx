import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Notification } from '../Notification';

describe('Notification', () => {
  const content = 'This is a test messgae';
  const { prefixCls } = Notification.defaultProps;

  it('renders basic Notification correctly', () => {
    const title = 'Test Title';
    const wrapper = shallow(
      <Notification title={title} content={content} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.text().indexOf(title)).not.toBe(-1);
    expect(wrapper.text().indexOf(content)).not.toBe(-1);
  });

  it('trigger event correctly', () => {
    const mouseEnter = jest.fn();
    const mouseLeave = jest.fn();
    const click = jest.fn();
    const wrapper = mount(
      <Notification
        content={content}
        onMouseEnter={mouseEnter}
        onMouseLeave={mouseLeave}
        onClick={click}
      />,
    );
    wrapper
      .find('div')
      .at(0)
      .simulate('mouseenter')
      .simulate('mouseleave');
    wrapper.find(`.${prefixCls}__content`).simulate('click');
    expect(mouseEnter).toHaveBeenCalledTimes(1);
    expect(mouseLeave).toHaveBeenCalledTimes(1);
    expect(click).toHaveBeenCalledTimes(1);
  });

  it('render custom icon correctly', () => {
    const wrapper = mount(
      <Notification content={content} icon={<i>?</i>} />,
    );
    expect(
      wrapper
        .find(`.${prefixCls}__icon`)
        .find('i')
        .text(),
    ).toBe('?');
  });

  it('render unknow icon correctly', () => {
    const wrapper = mount(
      <Notification content={content} icon="unknown-icon" />,
    );
    expect(wrapper.find(`.${prefixCls}__icon`)).toHaveLength(0);
  });

  it('render custom footer correctly', () => {
    const text = 'this is a custom footer';
    const wrapper = mount(
      <Notification content={content} footer={<div>{text}</div>} />,
    );
    expect(wrapper.find(`.${prefixCls}__foot`).text()).toBe(text);
  });
});
