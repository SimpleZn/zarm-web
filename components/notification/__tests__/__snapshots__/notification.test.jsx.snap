// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders basic Notification correctly 1`] = `
<div
  className="zw-notification"
  style={Object {}}
>
  <div
    className="zw-notification__content"
  >
    <div
      className="zw-notification__head"
    >
      Test Title
    </div>
    <div
      className="zw-notification__body"
    >
      This is a test messgae
    </div>
  </div>
  <Icon
    className="zw-notification__close"
    prefixCls="zw-icon"
    size="sm"
    theme="default"
    type="wrong"
    viewBox="0 0 32 32"
  />
</div>
`;
